File: packages/service/src/discord/discord-client.ts
import { Service } from 'typedi';
import { Client, GatewayIntentBits, Message } from 'discord.js';
import { Configuration } from '@service/infrastructure/configuration';
import { CommandExecutorService } from '@service/domain/commands/command-executor-service';
import { Logger } from '@service/infrastructure/logger';
import { UsersService } from '@service/domain/users/users-service';
import { ExecutionContext } from '@service/domain/commands/execution-context';
import { User } from '@service/domain/users/user';

const DISCORD_LOGIN_COMMAND_KEY = 'discord-login';

@Service()
class DiscordClient {
  private static logger: Logger = new Logger('discord-client');

  private client: Client;

  constructor(private configuration: Configuration, private commandExecutorService: CommandExecutorService, private usersService: UsersService) {
    this.client = new Client({
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
      ],
    });
    this.client.on('messageCreate', (message) => this.messageCreate(message));
  }

  public async login(): Promise<void> {
    if (!this.configuration.DISCORD_TOKEN) {
      DiscordClient.logger.error('Discord token was not provided');
      DiscordClient.logger.info('Running in local only mode');
      return;
    }

    await this.client.login(this.configuration.DISCORD_TOKEN);
    DiscordClient.logger.info('Logged in to Discord');
  }

  private async messageCreate(message: Message): Promise<void> {
    if (!message.content.startsWith(`${this.configuration.COMMAND_PROMPT} `)) return;
    if (message.channelId !== this.configuration.DISCORD_CHANNEL_ID) return;

    const discordId: string = message.author.id;
    const user: (User | null) = await this.usersService.getByDiscordId(discordId);
    const isLoginCommand: boolean = message.content.startsWith(`${this.configuration.COMMAND_PROMPT} ${DISCORD_LOGIN_COMMAND_KEY}`);
    const hasConnectedAccount: boolean = !!user;

    if (!hasConnectedAccount && !isLoginCommand) {
      await message.reply(
        `You have to connect your Discord account with LebkuchenFM\nUse \`${this.configuration.COMMAND_PROMPT} ${DISCORD_LOGIN_COMMAND_KEY} <lebkuchen-fm-username>\` to login.`,
      );
      return;
    }

    const context: ExecutionContext = {
      discordId,
      user: user!, // user might be null but this is only the case for discord-login command
    };

    const commandProcessingResponse = await this.commandExecutorService.processFromText(message.content, context);

    try {
      await message.reply(commandProcessingResponse.message.markdown);
    } catch (error) {
      DiscordClient.logger.error('There was a problem sending command response');
      DiscordClient.logger.withError(error as Error);
    }
  }
}

export { DiscordClient };

File: service/src/main/kotlin/xyz/lebkuchenfm/external/discord/DiscordClient.kt
package xyz.lebkuchenfm.external.discord

import dev.kord.common.entity.PresenceStatus
import dev.kord.core.Kord
import dev.kord.core.behavior.reply
import dev.kord.core.event.message.MessageCreateEvent
import dev.kord.gateway.Intent
import dev.kord.gateway.Intents
import dev.kord.gateway.PrivilegedIntent
import io.github.oshai.kotlinlogging.KotlinLogging
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.filterIsInstance
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import xyz.lebkuchenfm.domain.commands.CommandExecutorService
import xyz.lebkuchenfm.domain.commands.ExecutionContext
import xyz.lebkuchenfm.domain.integrations.DiscordIntegration
import xyz.lebkuchenfm.domain.users.UsersService

private val logger = KotlinLogging.logger {}

class DiscordClient(
    integration: DiscordIntegration?,
    private val commandExecutorService: CommandExecutorService,
    private val userService: UsersService,
) {
    private var token: String? = integration?.token
    private var commandPrompt: String? = integration?.effectiveCommandPrompt
    private var channelId: String? = integration?.channelId

    private var scope: CoroutineScope? = null
    private var kord: Kord? = null
    private var activeJob: Job? = null

    fun start(scope: CoroutineScope) {
        this.scope = scope
        activeJob = scope.launch { connect() }
    }

    suspend fun reconfigure(integration: DiscordIntegration?) {
        activeJob?.cancel()
        activeJob = null
        kord?.shutdown()
        kord = null

        token = integration?.token
        commandPrompt = integration?.effectiveCommandPrompt
        channelId = integration?.channelId

        scope?.let { start(it) }
    }

    private suspend fun connect() {
        val currentToken = token
        if (currentToken == null) {
            logger.warn { "Discord Bot token not provided." }
            return
        }

        val currentCommandPrompt = commandPrompt
        if (currentCommandPrompt == null) {
            logger.warn { "CommandPrompt prompt not provided." }
            return
        }

        val currentChannelId = channelId
        if (currentChannelId == null) {
            logger.warn { "Channel id not provided." }
            return
        }

        val newKord = Kord(currentToken)
        kord = newKord

        newKord.events
            .filterIsInstance<MessageCreateEvent>()
            .map { it.message }
            .filter { it.channelId.value.toString() == currentChannelId }
            .filter { it.content.split(' ').firstOrNull() == currentCommandPrompt }
            .filter { it.author != null }
            .filter { it.author?.isBot == false }
            .onEach {
                val author = requireNotNull(it.author)
                val user = userService.getByDiscordId(author.id.toString())

                when {
                    user == null -> {
                        it.reply { content = "You have to link your Discord account with LebkuchenFM user." }
                    }

                    user.secret == null -> {
                        it.reply { content = "You must login to LebkuchenFM, before you can use Discord integration." }
                    }

                    else -> {
                        val context = ExecutionContext(
                            username = user.data.name,
                            grantedScopes = user.effectiveScopes,
                            commandPrompt = currentCommandPrompt,
                        )
                        val commandText = it.content.replace("$currentCommandPrompt ", "")
                        val result = commandExecutorService.executeFromText(commandText, context)
                        it.reply { content = result.markdown }
                    }
                }
            }
            .launchIn(newKord)

        newKord.login {
            @OptIn(PrivilegedIntent::class)
            intents = Intents(
                Intent.Guilds,
                Intent.MessageContent,
                Intent.GuildMessages,
            )
            presence {
                status = PresenceStatus.Online
            }
        }
    }
}
